### PR TITLE
refactor(batch): Use futures-async-stream to implement NestedLoopJoin executor

### PR DESCRIPTION
**Pull request by D2Lark**
Signed-off-by: d2lark <lichengamoy@gmail.com>

## Implement NestedLoopJoin using Executor2 trait.

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Summarize your change (**mandatory**)
- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

## Refer to a related PR or issue link (optional)
close https://github.com/singularity-data/risingwave/issues/1948

### CODE REVIEW

File: src/batch/src/executor2/join/nested_loop_join.rs [hunk 1]
```diff
@@ -89,83 +91,67 @@ enum NestedLoopJoinState {
...
+        self.do_execute()
+    }
+}
 
+impl NestedLoopJoinExecutor2 {
+    #[try_stream(boxed, ok = DataChunk, error = RwError)]
+    async fn do_execute(mut self: Box<Self>) {
+        let mut state = NestedLoopJoinState::Build;
+        loop {
+            match state {
```
Comment (D2Lark) [inline src/batch/src/executor2/join/nested_loop_join.rs:113]: @liurenjie1024 I think the[ previous hash join implement](https://github.com/singularity-data/risingwave/pull/2119/files#r861413371)  is not very good, this change retains the match control flow is more intuitive, which one do you think is better

File: src/batch/src/executor2/join/nested_loop_join.rs [hunk 2]
```diff
@@ -89,83 +91,67 @@ enum NestedLoopJoinState {
...
+                        state = NestedLoopJoinState::Probe;
+                    }
                     if let Some(data_chunk) = ret {
-                        return Ok(Some(data_chunk));
+                        yield data_chunk;
                     }
                 }
                 NestedLoopJoinState::Probe => {
-                    let ret = self.probe(false).await?;
+                    let ret = self.probe(false, &mut state).await?;
```
Comment (BowenXiao1999) [inline src/batch/src/executor2/join/nested_loop_join.rs:128]: What's the reason not use state as self member fields like before?

**Review by BowenXiao1999 — COMMENTED** [review]
General LGTM

**Review by BowenXiao1999 — APPROVED** [review]

File: src/batch/src/executor2/join/nested_loop_join.rs [hunk 3]
```diff
@@ -89,83 +91,67 @@ enum NestedLoopJoinState {
...
+                    self.build_table.load_data().await?;
+                    state = NestedLoopJoinState::FirstProbe;
+                }
                 NestedLoopJoinState::FirstProbe => {
-                    let ret = self.probe(true).await?;
-                    self.state = NestedLoopJoinState::Probe;
+                    let ret = self.probe(true, &mut state).await?;
+                    if state == NestedLoopJoinState::FirstProbe {
+                        state = NestedLoopJoinState::Probe;
+                    }
```
Comment (BowenXiao1999) [inline src/batch/src/executor2/join/nested_loop_join.rs:122]: Is this check necessary? the state will be mutated to `Probe` in self.probe?

Comment (D2Lark) [reply src/batch/src/executor2/join/nested_loop_join.rs:122]: I should change the state in probe() instead :)

File: src/batch/src/executor2/join/nested_loop_join.rs (hunk 2 above)
Comment (D2Lark) [reply src/batch/src/executor2/join/nested_loop_join.rs:128]: I refer to this [pr](https://github.com/singularity-data/risingwave/pull/2119#discussion_r857426526) and consider that there is no need of a state machine to implement stream, so I put state as a local variable

**Comment by TennyZhuang** [comment]
CI is running: https://github.com/singularity-data/risingwave/actions/runs/2267874061

**Comment by TennyZhuang** [comment]
Thanks!

**Pull request merged**